### PR TITLE
TOF and dSpacing Units for GroupCalibrationWorkspace

### PR DESCRIPTION
**Pull request by darshdinger**
## Description of work
This defect was relatively straight forward once the issue was understood. The work time for this task was longer than expected due to my lack in understanding of the Calibration Workflow.

## Explanation of work
This consisted largely of adding some branching logic and adding an addition parameter to `convertAndFocusReturn` method within `GroupDiffractionCalibration`. The arguments 

## To test
Please run Calibration Workflow twice. Steps to follow for practical testing:

1. Run Calibration Workflow w/ these inputs:
```python
Run Number: 46680
Convergence Threshold: 0.1
Peak Intensity Threshold: 0.0001
Bins Across Peak Width: 10
Sample: Diamond_001.json
Grouping File: column lite
```

2. Check the workspace list to ensure the proper units are retained

## Link to EWM item

[EWM#3819](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=3819)

### CODE REVIEW

File: src/snapred/backend/recipe/algorithm/GroupDiffractionCalibration.py [hunk 1]
```diff
@@ -275,24 +276,30 @@ def PyExec(self) -> None:
             InstrumentWorkspace=self.wsTOF,
             CalibrationWorkspace=self.DIFCfinal,
         )
-        self.convertAndFocusAndReturn(self.wsTOF, self.outputWStof, "after")
+        self.convertAndFocusAndReturn(self.wsTOF, self.outputWStof, "after", self.units)
```
Comment (rboston628) [inline src/snapred/backend/recipe/algorithm/GroupDiffractionCalibration.py:279]: This should always call it with "dSpacing".  As per Malcolm, the output should just always be in d-spacing.

File: src/snapred/backend/dao/ingredients/DiffractionCalibrationIngredients.py [hunk 2]
```diff
@@ -16,3 +16,10 @@ class DiffractionCalibrationIngredients(BaseModel):
     groupedPeakLists: List[GroupPeakList]
     convergenceThreshold: float
     maxOffset: float = Config["calibration.diffraction.maximumOffset"]
+    units: str = "TOF"
+
+    @validator("units")
+    def validatedSpacing(cls, v):
+        if v not in ["TOF", "dSpacing"]:
+            raise ValueError(f"dSpacing must be either 'TOF' or 'dSpacing', not '{v}'")
+        return v
```
Comment (rboston628) [inline src/snapred/backend/dao/ingredients/DiffractionCalibrationIngredients.py:25]: For future reference, the simpler way to do this is with the `Literal` type, as in
``` python
units: Literal["TOF", "dSpacing"]
```
with no further validators needed.

Comment (darshdinger) [reply src/snapred/backend/dao/ingredients/DiffractionCalibrationIngredients.py:25]: Thanks for this tip! Guess I didn't think about doing it this way, but this is much better than creating validation logic.

File: src/snapred/backend/recipe/algorithm/GroupDiffractionCalibration.py (hunk 1 above)
Comment (darshdinger) [reply src/snapred/backend/recipe/algorithm/GroupDiffractionCalibration.py]: Okay, now I have a better understanding of the requirements of this story. I've been so used to hearing the word "flag" used in reference for propagating backend functionality to the frontend UI for user control but in this case, it just meant a simple extra input argument for the method making this task much simpler than I imagined.  Thank you for the heads up!

File: src/snapred/backend/recipe/algorithm/GroupDiffractionCalibration.py [hunk 3]
```diff
@@ -166,7 +167,7 @@ def unbagGroceries(self):
 
         # process and diffraction focus the input data
         # must convert to d-spacing, diffraction focus, ragged rebin, then convert back to TOF
-        self.convertAndFocusAndReturn(self.wsTOF, self.outputWStof, "before")
+        self.convertAndFocusAndReturn(self.wsTOF, self.outputWStof, "before", self.units)
```
[the remark this answers is not in the document]
Comment (darshdinger) [reply src/snapred/backend/recipe/algorithm/GroupDiffractionCalibration.py]: Done!

File: src/snapred/backend/recipe/algorithm/GroupDiffractionCalibration.py [hunk 4]
```diff
@@ -72,6 +72,7 @@ def chopIngredients(self, ingredients: Ingredients) -> None:
         self.dMin = ingredients.pixelGroup.dMin()
         self.dMax = ingredients.pixelGroup.dMax()
         self.dBin = ingredients.pixelGroup.dBin()
+        self.units = ingredients.units
```
Comment (rboston628) [inline src/snapred/backend/recipe/algorithm/GroupDiffractionCalibration.py:75]: remove

File: src/snapred/backend/dao/ingredients/DiffractionCalibrationIngredients.py [hunk 5]
```diff
@@ -1,6 +1,6 @@
 from typing import Dict, List
 
-from pydantic import BaseModel
+from pydantic import BaseModel, validator
```
Comment (rboston628) [inline src/snapred/backend/dao/ingredients/DiffractionCalibrationIngredients.py:3]: remove

**Comment by rboston628** [comment]
Everything ran with no issues using run 58882, and making sure the peak threshold was small enough to allow two peaks per group.  The output is in d-spacing, and has a name reflecting this fact.
image

**Review by rboston628 — APPROVED** [review]

**Pull request merged**